store/pg_engine.go

```diff
@@ -252,85 +252,74 @@ const (
-		latestDataPath := fmt.Sprintf("migration/%s/%s", db.schemaVersion.String(), latestDataFile)
+		latestDataPath := fmt.Sprintf("migration/%s/%s", cutoffSchemaVersion.String(), latestDataFile)
 		dataBuf, err := migrationFS.ReadFile(latestDataPath)
 		if err != nil {
-			return fmt.Errorf("failed to read latest data %q, error %w", latestSchemaPath, err)
+			return semver.Version{}, fmt.Errorf("failed to read latest data %q, error %w", latestSchemaPath, err)
 		}
 		// We will create the database together with initial schema and data migration.
 		stmt := fmt.Sprintf("CREATE DATABASE %s;\n\\connect \"%s\";\n%s\n%s", databaseName, databaseName, buf, dataBuf)
 		if _, _, err := d.ExecuteMigration(
 			ctx,
 			&dbdriver.MigrationInfo{
-				ReleaseVersion:        db.serverVersion,
+				ReleaseVersion:        serverVersion,
 				UseSemanticVersion:    true,
-				Version:               db.schemaVersion.String(),
+				Version:               cutoffSchemaVersion.String(),
 				SemanticVersionSuffix: time.Now().Format("20060102150405"),
 				Namespace:             databaseName,
 				Database:              databaseName,
 				Environment:           "", /* unused in execute migration */
 				Source:                dbdriver.LIBRARY,
 				Type:                  dbdriver.Migrate,
-				Description:           fmt.Sprintf("Initial migration version %s server version %s with file %s.", db.schemaVersion, db.serverVersion, latestSchemaPath),
+				Description:           fmt.Sprintf("Initial migration version %s server version %s with file %s.", cutoffSchemaVersion, serverVersion, latestSchemaPath),
 				CreateDatabase:        true,
 			},
 			stmt,
 		); err != nil {
-			return fmt.Errorf("failed to migrate initial schema version %q, error: %v", latestSchemaPath, err)
+			return semver.Version{}, fmt.Errorf("failed to migrate initial schema version %q, error: %v", latestSchemaPath, err)
 		}
-		db.l.Info("Completed database initial migration.")
-		return nil
+		l.Info(fmt.Sprintf("Completed database initial migration with version %s.", cutoffSchemaVersion))
+		return cutoffSchemaVersion, nil
 	}
 
 	// Apply migrations
-	versionNames, err := fs.Glob(migrationFS, "migration/*")
+	versions, err := getMigrationFileVersions()
 	if err != nil {
-		return err
+		return semver.Version{}, err
 	}
-	var versions []semver.Version
-	for _, name := range versionNames {
-		v, err := semver.Make(strings.TrimPrefix(name, "migration/"))
-		if err != nil {
-			return fmt.Errorf("invalid migration file path %q, error %w", name, err)
-		}
-		versions = append(versions, v)
-	}
-	// Sort the migration semantic version in ascending order.
-	sort.Slice(versions, func(i, j int) bool {
-		return versions[i].LT(versions[j])
-	})
 
-	migrateVersions, messages := getMigrationVersions(versions, db.schemaVersion, *curVer)
+	migrateVersions, messages := getMigrationVersions(versions, cutoffSchemaVersion, *curVer)
 	for _, message := range messages {
-		db.l.Info(message)
+		l.Info(message)
 	}
 
```

Comment:
If migrateVersions is empty, we can do an early return and add a log. Thus we also don't need the conditional if log at the end

Reply:
Done.